cpp/src/c_api/replicate_edgelist.cpp

```diff
@@ -18,111 +18,167 @@
 
-    auto gathered_edgelist_dsts = cugraph::c_api::detail::device_allgatherv(
-      handle, raft::device_span<vertex_t const>(edgelist_dsts.data(), edgelist_dsts.size()));
+    if (edgelist_type_ids) {
+      raft::copy<edge_type_id_t>(edgelist_type_ids->data(),
+                                 edge_type_ids_->as_type<edge_type_id_t>(),
+                                 edge_type_ids_->size_,
+                                 handle_.get_stream());
+    }
+
+    auto& comm                  = handle_.get_comms();
+    auto gathered_edgelist_srcs = cugraph::detail::device_allgatherv(
+      handle_, comm, raft::device_span<vertex_t const>(edgelist_srcs.data(), edgelist_srcs.size()));
 
-    rmm::device_uvector<size_t> edge_offsets(2, handle.get_stream());
+    auto gathered_edgelist_dsts = cugraph::detail::device_allgatherv(
+      handle_, comm, raft::device_span<vertex_t const>(edgelist_dsts.data(), edgelist_dsts.size()));
+
+    rmm::device_uvector<size_t> edge_offsets(2, handle_.get_stream());
     std::vector<size_t> h_edge_offsets{{0, gathered_edgelist_srcs.size()}};
     raft::update_device(
-      edge_offsets.data(), h_edge_offsets.data(), h_edge_offsets.size(), handle.get_stream());
+      edge_offsets.data(), h_edge_offsets.data(), h_edge_offsets.size(), handle_.get_stream());
 
-    // FIXME: Hnadle this case better
-    if (edgelist_weights) {
-      auto gathered_edgelist_weights = cugraph::c_api::detail::device_allgatherv(
-        handle,
-        raft::device_span<weight_t const>(edgelist_weights->data(), edgelist_weights->size()));
-
-      *result = new cugraph::c_api::cugraph_induced_subgraph_result_t{
-        new cugraph::c_api::cugraph_type_erased_device_array_t(gathered_edgelist_srcs, src->type_),
-        new cugraph::c_api::cugraph_type_erased_device_array_t(gathered_edgelist_dsts, dst->type_),
-        new cugraph::c_api::cugraph_type_erased_device_array_t(gathered_edgelist_weights,
-                                                               weights->type_),
-        new cugraph::c_api::cugraph_type_erased_device_array_t(edge_offsets,
-                                                               cugraph_data_type_id_t::SIZE_T)};
-    } else {
-      *result = new cugraph::c_api::cugraph_induced_subgraph_result_t{
-        new cugraph::c_api::cugraph_type_erased_device_array_t(gathered_edgelist_srcs, src->type_),
-        new cugraph::c_api::cugraph_type_erased_device_array_t(gathered_edgelist_dsts, dst->type_),
-        NULL,
-        new cugraph::c_api::cugraph_type_erased_device_array_t(edge_offsets,
-                                                               cugraph_data_type_id_t::SIZE_T)};
-    }
+    cugraph::c_api::cugraph_induced_subgraph_result_t* result = NULL;
 
-  } catch (std::exception const& ex) {
-    *error = new cugraph::c_api::cugraph_error_t{ex.what()};
-    return CUGRAPH_UNKNOWN_ERROR;
-  }
+    std::optional<rmm::device_uvector<weight_t>> gathered_weights =
```

Comment:
If you overwrite this can become something like:

```
if (edgelist_weights) {
   edgelist_weights = cugraph::detail::device_allgatherv(handle_, comm, raft::device_span<weight_t const>(edgelist_weights->data(), edgelist_weights->size()));
}
```

Which is a bit easier to read.